server/index.js

```diff
@@ -65,17 +65,45 @@ app.post('/marvel/registration', (req, res, next) => {
   argon2
     .hash(password)
     .then((passwordHash) => {
+      const usernameSql = `
+        select "id", "username", "email", "createdAt" from "users" where "username" = $1 limit 1
+      `;
+      const usernameParams = [username];
+
+      const emailSql = `
+        select "id", "username", "email", "createdAt" from "users" where "email" = $1 limit 1
+      `;
+      const emailParams = [email];
+
+      return Promise.all([
+        db.query(usernameSql, usernameParams),
+        db.query(emailSql, emailParams),
+        argon2.hash(password)
+      ]);
+    })
+    .then(([usernameResult, emailResult, passwordHash]) => {
+      if (usernameResult.rows.length > 0) {
+        throw new ClientError(409, 'username already exists');
+      }
+      if (emailResult.rows.length > 0) {
+        throw new ClientError(409, 'email already exists');
+      }
       const sql = `
         insert into "users" ("username", "passwordHash", "email")
         values ($1, $2, $3)
+        on conflict do nothing
         returning "id", "username", "email", "createdAt"
       `;
       const params = [username, passwordHash, email];
       return db.query(sql, params);
     })
     .then((result) => {
-      const [user] = result.rows;
-      res.status(201).json(user);
+      if (result.rowCount === 0) {
+        throw new ClientError(409, 'username or email already exists');
+      } else {
+        const [user] = result.rows;
+        res.status(201).json(user);
+      }
```

Comment:
All this code would look much simpler with `async/await`. You should make that transition after you've got your features finished.

Reply:
Will do!